File: 2021/day12/cavePaths.dart
import 'dart:io';

void main() async {
    final conns = await getInput("sample.txt");
    print('Contents: ${conns}');
}

Future<Map<String,Set<String>>> getInput(filename) async {
  final file = File(filename);
  final lines = await file.readAsString();
  var conns = Map<String,Set<String>>();
  conns["start"] = Set();
  lines.split("\n")
    .map((line) => line.split("-"))
    .forEach((pair) {
      var p1,p2 = pair;
      if(conns.containsKey(p1)) {
        p2.forEach((e) {
          conns[p1].add(e);
        });
      } else {
        conns[p1] = Set<String>();
      }
    });
  
  return conns;
}

File: 2021/day12/cavePaths.jl
function remove(col, item)
    return filter(x -> x != item ,col)[1]
end

function getInput(filename)
    lines = filename |> open |> readlines
    connections = Dict(
        "start" => Set()
    )
    for line in lines
        nodes = split(line, "-")
        if "start" in nodes
            other = remove(nodes, "start")
            push!(connections["start"], other)
        elseif "end" in nodes
            other = remove(nodes, "end")
            if !(other in keys(connections))
                connections[other] = Set()
            end
            push!(connections[other], "end")
        else
            for n in nodes
                other = remove(nodes, n)
                if !(n in keys(connections))
                    connections[n] = Set()
                end
                push!(connections[n], other)
            end
        end
    end
    return connections
end

function isAllUpper(s)
    return reduce(&, map(c -> isuppercase(c), collect.(s)))
end

function nextSteps(conns, path)
    lastNode = last(path)
    steps = filter( s -> isAllUpper(s) || !(s in path), conns[lastNode])
    newPaths = []
    for s in steps
        push!(newPaths, push!(path, s))
        pop!(path)
    end
    return newPaths
end

function enumeratePaths(conns, paths, finishedPaths)
    if isempty(paths)
        return finishedPaths
    else
        newPaths = Set()
        newFinished = union(finishedPaths, filter(x -> "end" in x, newPaths))
        enumeratePaths(conns, newPaths, newFinished)
    end
end

conns = getInput("sample.txt")